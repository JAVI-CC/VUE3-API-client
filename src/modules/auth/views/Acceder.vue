<template>
  <div class="q-mt-xl"></div>
  <div class="div-acceder q-py-lg q-px-lg text-center">
    <div class="text-h4 text-weight-light q-mb-lg">Iniciar sesión</div>
    <q-form data-testid="form" ref="formAcceder" @submit.prevent="formLogin">
      <div class="col-md-12 q-mb-xs">
        <q-input
          clearable
          color="white"
          name="email"
          label="Email"
          data-cy="acceder-input-email"
          v-model="email"
          :rules="[
            (val) => v$.email.required.$response || 'Debes rellenar este campo',
            (val) => v$.email.email.$response || 'Debe ser un email',
          ]"
        >
          <template v-slot:prepend>
            <q-icon name="email" />
          </template>
        </q-input>
      </div>

      <div class="col-md-12 q-mb-md">
        <q-input
          color="white"
          name="contraseña"
          clearable
          label="Contraseña"
          data-cy="acceder-input-password"
          v-model="password"
          :type="pwdTextShow ? 'password' : 'text'"
          :rules="[
            (val) => v$.password.required.$response || 'Debes rellenar este campo',
            (val) =>
              v$.password.minLength.$response || 'Debe de tener al menos 8 caracteres',
            (val) =>
              v$.password.maxLength.$response || 'Debe de tener menos de 100 caracteres',
          ]"
        >
          <template v-slot:prepend>
            <q-icon name="password" />
          </template>
          <template v-slot:append>
            <q-icon
              :name="pwdTextShow ? 'visibility_off' : 'visibility'"
              class="cursor-pointer"
              @click="pwdTextShow = !pwdTextShow"
            />
          </template>
        </q-input>
      </div>

      <div class="col-md-12">
        <q-btn
          no-caps
          :disabled="v$.email.$invalid || v$.password.$invalid"
          color="primary"
          rounded
          icon-right="lock_outline"
          label="Acceder"
          data-cy="acceder-button-submit"
          type="submit"
        />
      </div>
    </q-form>
  </div>
  <div class="div-sin-cuenta q-py-md q-px-lg text-center">
      <div class="text-h6 q-mb-md">¿ No tienes cuenta ?</div>
       <q-btn data-cy="acceder-button-crear-cuenta" outline class="text-h6" :class="dark ? 'text-white' : 'text-black'" padding="sm xl" no-caps label="Crear cuenta" @click="toRegister" />
  </div>
  <modal-info @confirm="closeModal" :open="modalError">
     <q-card-section>{{modalMessageError}}</q-card-section>
  </modal-info>
</template>

<script>
import { defineAsyncComponent, ref } from "vue";
import useAuth from "../composables/useAuth";
import useLayout from "src/modules/layout/composables/useLayout";
import { required, email, minLength, maxLength } from "@vuelidate/validators";
import useVuelidate from "@vuelidate/core/dist/index.es";
export default {
  name: "Acceder",
    components: {
    ModalInfo: defineAsyncComponent(() => import('/src/modules/layout/components/Modal.vue')),
  },
  setup() {
    const { dark } = useLayout();
    const { login, inicio, notifLogin, toRegister, } = useAuth();
    const email = ref("");
    const password = ref("");
    const pwdTextShow = ref(true);
    const modalError = ref(false)
    const modalMessageError = ref("") 
    const formAcceder = ref("")
    return {
      //Methods
      closeModal(){
        modalError.value = false
      },
      formLogin: async() => {
        const logged = await login(email.value, password.value)
        if(logged != true) {
          modalError.value = true
          modalMessageError.value = logged.error
        } else {
           notifLogin()
           inicio()
           email.value = ""
           password.value = ""
          formAcceder.value.reset();
        }
      },
      toRegister,

      //Variables
      dark,
      email,
      formAcceder,
      modalError,
      modalMessageError,
      password,
      pwdTextShow,
      v$: useVuelidate(),
    };
  },
  validations: {
    email: {
      required,
      email,
    },
    password: {
      required,
      minLength: minLength(8),
      maxLength: maxLength(100),
    },
  },
};
</script>

<style lang="sass" scoped>
.div-acceder
  width: 325px
  margin: 0 auto
  border: 1px solid gray
  border-radius: 5px

.div-sin-cuenta
  width: 325px
  margin: 0 auto
</style>
