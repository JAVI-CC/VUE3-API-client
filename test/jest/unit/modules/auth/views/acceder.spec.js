import { shallowMount } from '@vue/test-utils'
import Acceder from 'src/modules/auth/views/Acceder.vue'
import createVuexStore from "../../../mock-data/mock-store";
import { Notify, QBtn } from 'quasar'

const mockRoute = { fullPath: jest.fn() }
const mockRouter = { push: jest.fn() }

jest.mock('vue-router', () => ({
    useRoute: () => mockRoute,
    useRouter: () => mockRouter
}))

Notify.create = () => {
    return jest.fn()
}

describe('Pruebas en el Acceder Component', () => {

    let wrapper
    const store = createVuexStore({
        logged: false,
        user: null,
    })

    store.dispatch = jest.fn()

    beforeEach(() =>
        wrapper = shallowMount(Acceder, {
            global: {
                plugins: [store]
            }
        }),
        jest.clearAllMocks())

    test('debe de hacer match con el snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('Titulo debe de tener el valor por defecto "Iniciar sesi贸n"', () => {
        const titulo = wrapper.find('.text-h4')
        expect(titulo.text()).toBe('Iniciar sesi贸n')
    })

    test('credenciales incorrectas, disparar el Modal dialog', async () => {

        store.dispatch.mockReturnValueOnce({ error: "Las credenciales no s贸n correctas" })
        await wrapper.find('[data-testid="form"]').trigger('submit')
        expect(store.dispatch).toHaveBeenCalledWith("auth/login", { email: "", password: "" })
        expect(wrapper.vm.modalError).toBeTruthy()
        expect(wrapper.vm.modalMessageError).toBe('Las credenciales no s贸n correctas')

        //Comprobar que el componente del modal-info exista
        const modalInfo = wrapper.find('modal-info-stub')
        expect(modalInfo.exists()).toBeTruthy()
        expect(modalInfo.attributes('open')).toBe("true")
    })

    test.skip('credenciales correctas, redirige juegos', async () => {
        store.dispatch.mockReturnValue(true)
        wrapper.vm.email = 'test@email.com'
        wrapper.vm.password = '12345678'

        await wrapper.find('[data-testid="form"]').trigger('submit')
        expect(store.dispatch).toHaveBeenCalledWith("auth/login", { email: "test@email.com", password: "12345678" })
        expect(wrapper.vm.modalError).toBeFalsy()
        expect(wrapper.vm.email).toBe("")
        expect(wrapper.vm.password).toBe("")

        expect(mockRouter.push).toHaveBeenCalledWith({ "name": "juegos" });

        //Comprobar que el componente del modal-info exista
        const modalInfo = wrapper.find('modal-info-stub')
        expect(modalInfo.exists()).toBeTruthy()
        expect(modalInfo.attributes('open')).toBe("false")

    })

    test('clicar boton crear cuenta, redirige a registrar', async () => {

        const button = wrapper.findComponent(QBtn);
        await button.trigger('click');
        expect(mockRouter.push).toHaveBeenCalledWith({ "name": "registrar" });
    })

})