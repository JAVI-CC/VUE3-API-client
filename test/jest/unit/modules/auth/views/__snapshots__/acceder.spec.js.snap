// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el Acceder Component debe de hacer match con el snapshot 1`] = `
<div class="q-mt-xl"></div>
<div class="div-acceder q-py-lg q-px-lg text-center">
  <div class="text-h4 text-weight-light q-mb-lg">Iniciar sesión</div>
  <q-form-stub autofocus="false" noerrorfocus="false" noresetfocus="false" greedy="false" data-testid="form"></q-form-stub>
</div>
<div class="div-sin-cuenta q-py-md q-px-lg text-center">
  <div class="text-h6 q-mb-md">¿ No tienes cuenta ?</div>
  <q-btn-stub padding="sm xl" label="Crear cuenta" replace="false" exact="false" activeclass="q-router-link--active" exactactiveclass="q-router-link--exact-active" disable="false" type="button" round="false" outline="true" flat="false" unelevated="false" rounded="false" push="false" glossy="false" fab="false" fabmini="false" nocaps="true" nowrap="false" dense="false" ripple="true" align="center" stack="false" stretch="false" darkpercentage="false" data-cy="acceder-button-crear-cuenta" class="text-h6 text-white"></q-btn-stub>
</div>
<modal-info-stub open="false"></modal-info-stub>
`;
